File: win-prog/ProgWinVB/Buttons and Labels and Scrolls (Oh, My!)/CustomCheckBox/CustomCheckBox.vb
'-----------------------------------------------
' CustomCheckBox.vb (c) 2002 by Charles Petzold
'-----------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class CustomCheckBox
    Inherits Form

    Shared Sub Main()
        Application.Run(New CustomCheckBox())
    End Sub

    Sub New()
        Text = "Custom CheckBox Demo"

        Dim cyText As Integer = Font.Height
        Dim cxText As Integer = cyText \ 2
        Dim afs As FontStyle() = {FontStyle.Bold, FontStyle.Italic, _
                                  FontStyle.Underline, FontStyle.Strikeout}

        Dim lbl As New Label()
        lbl.Parent = Me
        lbl.Text = Text & ": Sample Text"
        lbl.AutoSize = True

        Dim i As Integer
        For i = 0 To 3
            Dim chkbox As New FontStyleCheckBox()
            chkbox.Parent = Me
            chkbox.Text = afs(i).ToString()
            chkbox.fntstyle = afs(i)
            chkbox.Location = New Point(2 * cxText, _
                                       (4 + 3 * i) * cyText \ 2)
            chkbox.Size = New Size(12 * cxText, cyText)
            AddHandler chkbox.CheckedChanged, _
                                    AddressOf CheckBoxOnCheckedChanged
        Next i
    End Sub

    Private Sub CheckBoxOnCheckedChanged(ByVal obj As Object, _
                                         ByVal ea As EventArgs)
        Dim fs As FontStyle = 0
        Dim lbl As Label = Nothing
        Dim ctrl As Control

        For Each ctrl In Controls
            If ctrl.GetType() Is GetType(Label) Then
                lbl = DirectCast(ctrl, Label)
            ElseIf ctrl.GetType() Is GetType(FontStyleCheckBox) Then
                If DirectCast(ctrl, FontStyleCheckBox).Checked Then
                    fs = fs Or DirectCast(ctrl, FontStyleCheckBox).fntstyle
                End If
            End If
        Next

        lbl.Font = New Font(lbl.Font, fs)
    End Sub
End Class

Class FontStyleCheckBox
    Inherits CheckBox

    Public fntstyle As FontStyle
End Class

File: win-prog/ProgWinCS/Buttons and Labels and Scrolls (Oh, My!)/CustomCheckBox/CustomCheckBox.cs
//---------------------------------------------
// CustomCheckBox.cs © 2001 by Charles Petzold
//---------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class CustomCheckBox: Form
{
     public static void Main()
     {
          Application.Run(new CustomCheckBox());
     }
     public CustomCheckBox()
     {
          Text = "Custom CheckBox Demo";

          int      cyText = Font.Height;
          int      cxText = cyText / 2;
          FontStyle[] afs = { FontStyle.Bold,      FontStyle.Italic, 
                              FontStyle.Underline, FontStyle.Strikeout };

          Label label    = new Label();
          label.Parent   = this;
          label.Text     = Text + ": Sample Text";
          label.AutoSize = true;

          for (int i = 0; i < 4; i++)
          {
               FontStyleCheckBox chkbox = new FontStyleCheckBox();
               chkbox.Parent = this;
               chkbox.Text = afs[i].ToString();
               chkbox.fontstyle = afs[i];
               chkbox.Location = new Point(2 * cxText, 
                                               (4 + 3 * i) * cyText / 2);
               chkbox.Size = new Size(12 * cxText, cyText);
               chkbox.CheckedChanged += 
                              new EventHandler(CheckBoxOnCheckedChanged);
          }
     }
     void CheckBoxOnCheckedChanged(object obj, EventArgs ea)
     {
          FontStyle fs = 0;
          Label     label = null;

          for (int i = 0; i < Controls.Count; i++)
          {
               Control ctrl = Controls[i];

               if (ctrl.GetType() == typeof(Label))
                    label = (Label) ctrl;

               else if (ctrl.GetType() == typeof(FontStyleCheckBox))
                    if (((FontStyleCheckBox) ctrl).Checked)
                         fs |= ((FontStyleCheckBox) ctrl).fontstyle;
          }
          label.Font = new Font(label.Font, fs);
     }
}
class FontStyleCheckBox: CheckBox
{
     public FontStyle fontstyle;
}